### ISSUE TITLE
[Bug] Empty nodes are not preserved

### ISSUE THREAD

**Issue by wintercounter** [Milkdown/milkdown#1579]
### Initial checklist

- [X] I agree to follow the [code of conduct](https://github.com/Saul-Mirone/milkdown/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched [issues](hhttps://github.com/Saul-Mirone/milkdown/issues) and [discussions](https://github.com/Saul-Mirone/milkdown/discussions) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

7.5.0

### Link to runnable example

https://milkdown.dev/playground?text=DISwdgpgBAjAUAupJQExyA%3D%3D

### Steps to reproduce

1. Add an empty line to Milkdown editor
2. Observe that Markdown in Codemirror gets updated correctly.
3. Make an edit in the markdown content.
4. Empty node gets removed.

Currently, when users edit their posts, they are losing the spacing between paragraphs because of this.

The markdown that gets saved into the database.
```
Line 1



Line 2
```

Upon edit we load the markdown into Milkdown, using (same code as in the playground)
```
const editor = get();
editor?.action((ctx) => {
  const view = ctx.get(editorViewCtx);
  const parser = ctx.get(parserCtx);
  const doc = parser(markdown);
  if (!doc) return;
  const state = view.state;
  view.dispatch(
    state.tr.replace(
      0,
      state.doc.content.size,
      new Slice(doc.content, 0, 0),
    ),
  );
});
```

Content becomes
```
Line 1

Line 2
```

Basically if I load again Milkdown's previous output, I won't get the same results again.

### Expected behavior

Empty nodes are preserved by the parser.

### Actual behavior

Empty node gets removed.

### Runtime

Chrome

### OS

Windows

### Build and bundle tools

Next.js

**Comment by shenzhongkang** [comment]
face the same issue. @Saul-Mirone

**Comment by insertmike** [comment]
any updates @Saul-Mirone ?

**Comment by Saul-Mirone** [comment]
@insertmike @shenzhongkang 
Nope. I investigate it a little but don't have enough time to work on this feature. So it's kind of on the roadmap. If you'd like to see it prioritized, contributions or donations can help us accelerate the development. I truly appreciate your support.

**Comment by insertmike** [comment]
Hold my beer @Saul-Mirone , let me look into this

**Comment by Saul-Mirone** [comment]
> Hold my beer [@Saul-Mirone](https://github.com/Saul-Mirone) , let me look into this

Thanks, the basic strategy is to convert empty paragraph node to a line break node (<br>). And transform the line break node to empty paragraph when parsing markdown.

**Comment by insertmike** [comment]
@Saul-Mirone Have a look, not the best, but found the culprit and opened a PR to kick things off: https://github.com/Milkdown/milkdown/pull/1758

**Comment by wintercounter** [comment]
@insertmike I believe this won't work well and it'll be a breaking change. We have users who were adding multiple linebreaks to their posts, which we wanted to prevent, so we're replacing 3 or more consecutive line breaks with only 2. This change would break our regexp, and possibly other real world apps too.

We're using `react-markdown` to render the posts, and in such case it is inserting empty `p` tags. `react-markdown` is also based on `remark`, maybe it's worth taking a look what they are doing differently.

**Comment by Saul-Mirone** [comment]
Implemented in https://github.com/Milkdown/milkdown/pull/1765

**Comment by wintercounter** [comment]
@Saul-Mirone I just added the plugin to our editor, but it's not the expected outcome.

I add 1 single enter after a word and the result is this:

```
Test\n\n<br />\n
```

First of all, I'm pretty sure it isn't supposed to have that many `\n`.

Also having hardcoded `<br />` wasn't the goal here at all, but for Milkdown to display things like `\n\n\n\n` correctly when previous content is loaded back in the editor.

The issue still stands. My user's previous posts that were created previously are still loading back wrong into Milkdown.

We don't allow any html tags in our rendered, so now all posts are displaying `<br />`. I had to remove the plugin and I had to add a regexp filter to our post render to "fix" our posts.

Please re-open the ticket.

**Comment by Saul-Mirone** [comment]
@wintercounter I don't think support `\n\n\n\n` is the goal because it won't work for other standard markdown platforms. For example, github readme, also `react-markdown` you mentioned above. It will cause the content looks different between milkdown and other platforms. `<br />` is supported. For your senario, I think you don't really need this plugin and you can use the preprocessor solution since it works well.

**Comment by wintercounter** [comment]
`react-markdown` will add empty paragraph in such case. If I'm unable to correctly load back `\n\n\n\n` into Milkdown then it shouldn't produce such output in the first place.

**Comment by Saul-Mirone** [comment]
<img width="1450" alt="Image" src="https://github.com/user-attachments/assets/cb0f85c7-efe8-4000-9a62-57e2b06aa2fd" />

I tried and seems they won't according to the playground. Could you please provide an example?

**Comment by wintercounter** [comment]
@Saul-Mirone  I tried to create a repro, which is when I realized that our styling of the editor was inconsistent compared to the markdown renderer. In the editor the team switched from flex to block (with flex there is no blinking caret when it's empty), but they didn't account for the gap flex applied by default. This means users were creating space between lines using multiple line-breaks. When they tried to edit, the line-breaks collapsed (correctly), and it looked like there was no gap between the paragraphs anymore. TLDR: this was an issue on out side.
